File: frontend/src/components/benchmark/content/degree_centrality/implementations/degree_centrality.java
import java.util.*;

public class DegreeCentrality {
    private Map<Integer, Set<Integer>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    public void addEdge(int u, int v) {
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.putIfAbsent(v, new HashSet<>());
        adjacency.get(u).add(v);
        adjacency.get(v).add(u);
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public int degreeCentrality(int node) {
        if (!adjacency.containsKey(node)) {
            return 0;
        }
        return adjacency.get(node).size();
    }

    public Map<Integer, Integer> allDegreeCentralities() {
        Map<Integer, Integer> result = new HashMap<>();
        for (int node : nodes) {
            result.put(node, degreeCentrality(node));
        }
        return result;
    }

    public List<Map.Entry<Integer, Integer>> topKCentralNodes(int k) {
        Map<Integer, Integer> centralities = allDegreeCentralities();
        List<Map.Entry<Integer, Integer>> result = new ArrayList<>(centralities.entrySet());
        
        result.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        
        return result.subList(0, Math.min(k, result.size()));
    }

    public double normalizedDegreeCentrality(int node) {
        if (!adjacency.containsKey(node) || nodes.size() <= 1) {
            return 0.0;
        }
        return (double) degreeCentrality(node) / (nodes.size() - 1);
    }

    public Map<Integer, Double> allNormalizedDegreeCentralities() {
        Map<Integer, Double> result = new HashMap<>();
        for (int node : nodes) {
            result.put(node, normalizedDegreeCentrality(node));
        }
        return result;
    }
}

File: frontend/src/components/benchmark/content/degree_centrality/implementations/degree_centrality.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

class DegreeCentrality {
private:
    unordered_map<int, unordered_set<int>> adjacency;
    unordered_set<int> nodes;

public:
    void add_edge(int u, int v) {
        adjacency[u].insert(v);
        adjacency[v].insert(u);
        nodes.insert(u);
        nodes.insert(v);
    }

    void build_from_edges(const vector<pair<int, int>>& edges) {
        for (const auto& edge : edges) {
            add_edge(edge.first, edge.second);
        }
    }

    int degree_centrality(int node) {
        if (adjacency.find(node) == adjacency.end()) {
            return 0;
        }
        return adjacency[node].size();
    }

    unordered_map<int, int> all_degree_centralities() {
        unordered_map<int, int> result;
        for (int node : nodes) {
            result[node] = degree_centrality(node);
        }
        return result;
    }

    vector<pair<int, int>> top_k_central_nodes(int k = 10) {
        auto centralities = all_degree_centralities();
        vector<pair<int, int>> result(centralities.begin(), centralities.end());
        
        sort(result.begin(), result.end(),
             [](const auto& a, const auto& b) {
                 return a.second > b.second;
             });
        
        if (k > (int)result.size()) {
            k = result.size();
        }
        
        return vector<pair<int, int>>(result.begin(), result.begin() + k);
    }

    double normalized_degree_centrality(int node) {
        if (adjacency.find(node) == adjacency.end() || nodes.size() <= 1) {
            return 0.0;
        }
        return (double)degree_centrality(node) / (nodes.size() - 1);
    }

    unordered_map<int, double> all_normalized_degree_centralities() {
        unordered_map<int, double> result;
        for (int node : nodes) {
            result[node] = normalized_degree_centrality(node);
        }
        return result;
    }
};